src/form.js

```diff
@@ -274,23 +276,63 @@ define([
          * @returns - The `id` of the added or already-existing instance.
          */
         addInstanceIfNotExists: function (attrs, mug, property) {
-            if (attrs.src === null || _.isUndefined(attrs.src)) {
-                // no ref counting for instances without a `src` since they cannot be dropped
-                this.instanceMetadata.push(InstanceMetadata({
-                    src: attrs.src,
-                    id: attrs.id
-                }));
-                return attrs.id;
+            function getUniqueId(id, ids) {
+                var temp = (id || "data") + "-",
+                    num = 1;
+                while (!id || ids.hasOwnProperty(id)) {
+                    id = temp + num;
+                    num++;
+                }
+                return id;
             }
-
-            var meta = _.find(this.instanceMetadata, function (m) {
+            var meta;
+            if (attrs.src) {
+                meta = _.find(this.instanceMetadata, function (m) {
                     return m.attributes.src === attrs.src;
                 });
+                if (!meta) {
+                    // attrs.src not found, try to find by id
+                    var ids = _.indexBy(this.instanceMetadata, function (m) {
+                            return m.attributes.id;
+                        });
+                    meta = attrs.id && ids.hasOwnProperty(attrs.id) ? ids[attrs.id] : null;
```

Comment:
Because `ids` is used again below by `getUniqueId(attrs.id, ids)`, which can do multiple lookups. I suppose that's a sufficiently far-out edge case that `_.find` wouldn't be too terrible.